nixos/modules/security/systemd-confinement.nix

```diff
@@ -98,8 +98,9 @@ in {
         wantsAPIVFS = lib.mkDefault (config.confinement.mode == "full-apivfs");
       in lib.mkIf config.confinement.enable {
         serviceConfig = {
-          RootDirectory = "/var/empty";
-          TemporaryFileSystem = "/";
+          RuntimeDirectory = [ "confinement/${mkPathSafeName name}" ];
+          RootDirectory = lib.mkForce "/run/confinement/${mkPathSafeName name}";
+          InaccessiblePaths = [ "-+/run/confinement/${mkPathSafeName name}" ];
```

Comment:
Urgh... this sucks :unamused: - we should probably add a subtest to make sure that stuff in `/run` also is as minimal as possible and only contains `/run/systemd/incoming`.

Reply:
Some services (`tor`, `transmission`, `gonic`, `navidrome`, `biboumi`, `pyload`, `tetrd`, `endlessh`, `shiori`, `freeciv`) are already using a `RuntimeDirectory=` as `RootDirectory=`, `mkForce` is here to switch to `systemd-confinement`'s root.

Reply:
> `mkForce` is here to switch to `systemd-confinement`'s root.

The issue I see with this is that it makes it non-obvious. If I set `RootDirectory` for a service and get a different one, I'd be very much confused. Without the `mkForce` I'd get an evaluation error that informs me about conflicting values.

So what's the whole motivation behind this? Is it so that it's easier to use `systemd-confinement` downstream?

Reply:
So far `systemd-confinement` is [never enabled by default in NixOS](https://discourse.nixos.org/t/nixos-policy-regarding-systemd-confinement/18976), and currently uses `TemporaryFileSystem = "/"` which undermines `RootDirectory=`.
@martinetd and I considered using a `mkOverride`, but `mkForce` being much more often used in `nixos/`, I've kept using it, to indeed make it possible to enable `systemd-confinement` on services which already have a transient `RootDirectory=`, as I don't see the point of using `systemd-confinement` if that's to use anything else than a transient `RootDirectory=`.
On second thought, reusing any transient `RootDirectory=` already set by the service would achieve the same, so I propose to switch to `mkDefault`.

Reply:
Hmm, I had a look at tor and I'm not sure it's going to work so well with confinement anyway? They're adding /etc and /nix/store as BindReadOnlyPaths in their own RootDirectory, so whatever we're binding more precisely won't make the rest of the store go away (didn't try to confine the service but we're not resting existing values as far as I can see)

(They're also setting PrivateTmp so they have a writable tmp directory in there, I'd be surprised if tor actually needs it.. It used to be conditioned to its controlSocket, but that is created in the run dir so woulld probably work without?)

So with that in mind I'd say such services are probably a lost battle, but I also agree it doesn't matter if we use our confinement rootdirectory or the one they set, so `mkDefault` sounds good to me.